File: leetcode/python/000036.valid-sudoku.py
from collections import defaultdict


class Solution:
	# Time complexity: O(9 ^ 2)
	# Space complexity: O(9 ^ 2)
	def isValidSudoku(self, board: list[list[str]]) -> bool:
		rows = defaultdict(set)
		cols = defaultdict(set)
		blocks = defaultdict(set) # key = (row // 3, col // 3)

		for row in range(len(board)):
			for col in range(len(board)):
				cur = board[row][col]
				if cur == ".":
					continue
				if (cur in rows[row] or
					cur in cols[col] or
					cur in blocks[(row // 3, col // 3)]):
					return False
				rows[row].add(cur)
				cols[col].add(cur)
				blocks[(row // 3, col // 3)].add(cur)

		return True

File: leetcode/javascript/000036.valid-sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 * Time complexity: O(n^2)
 * Space complexity: O(n^2)
 */
var isValidSudoku = function (board) {
  const rows = new SetMap();
  const cols = new SetMap();
  const blocks = new SetMap(); // key = `${row // 3},${col // 3}`

  for (let r = 0; r < board.length; r++) {
    for (let c = 0; c < board[0].length; c++) {
      const n = board[r][c];

      if (n === ".") continue;

      if (
        rows.get(r).has(n) || cols.get(c).has(n) ||
        blocks.get(`${Math.floor(r / 3)},${Math.floor(c / 3)}`).has(n)
      ) {
        return false;
      }

      rows.get(r).add(n);
      cols.get(c).add(n);
      blocks.get(`${Math.floor(r / 3)},${Math.floor(c / 3)}`).add(n);
    }
  }

  return true;
};

class SetMap {
  constructor() {
    this.map = new Map();
  }

  get(key) {
    if (!this.map.has(key)) {
      this.map.set(key, new Set());
    }
    return this.map.get(key);
  }

  set(key, value) {
    this.map.set(key, value);
  }
}
